onnxruntime/test/contrib_ops/optional_ops_test.cc

```diff
@@ -0,0 +1,145 @@
+      : OpTester("Optional", 1, kMSDomain) {}
+
+ protected:
+  void AddNodes(onnxruntime::Graph& graph,
+                std::vector<onnxruntime::NodeArg*>& graph_input_defs,
+                std::vector<onnxruntime::NodeArg*>& graph_output_defs,
+                std::vector<std::function<void(onnxruntime::Node& node)>>& /*add_attribute_funcs*/) override {
+    /*
+    * Both the Optional* ops produce intermediate outputs
+    
+                      Graph Input  
+                          |
+                          |
+                      Optional
+                          |
+                          |
+                      OptionalGetElement
+                         |
+                         |
+                       SampleOp  (use something other than Identity because it could be optimized away during Identity elimination)
+                                 Also preferably use something from kMsDomain as it seems like the setup to do cross-domain model creation
+                                 is not complete in OpTester.
+                                 SampleOp is just an Identity behavior-wise. 
+                         |
+                         | 
+                      Graph output
+      */
+
+    onnx::TypeProto tensor_type_proto;
+    tensor_type_proto.mutable_tensor_type()
+        ->set_elem_type(onnx::TensorProto_DataType::TensorProto_DataType_FLOAT);
+
+    onnx::TypeProto optional_type_proto;
+    optional_type_proto.mutable_optional_type()->mutable_elem_type()->CopyFrom(tensor_type_proto);
+
+    auto& optional_node_arg = graph.GetOrCreateNodeArg("optional_output", &optional_type_proto);
+    ORT_IGNORE_RETURN_VALUE(graph.AddNode("optional_create", "Optional", "Create optional type",
+                                          {graph_input_defs[0]}, {&optional_node_arg},
+                                          nullptr, kMSDomain));
+
+    auto& tensor_node_arg = graph.GetOrCreateNodeArg("tensor_output", &tensor_type_proto);
+    ORT_IGNORE_RETURN_VALUE(graph.AddNode("optional_get_element", "OptionalGetElement",
+                                          "Parse optional type",
+                                          {&optional_node_arg}, {&tensor_node_arg},
+                                          nullptr, kMSDomain));
+
+    ORT_IGNORE_RETURN_VALUE(graph.AddNode("SampleOp", "SampleOp",
+                                          "Identity",
+                                          {&tensor_node_arg}, {graph_output_defs[0]},
+                                          nullptr, kMSDomain));
+  }
+};
+
+TEST(OptionalOpTest, OptionalOpsValidateOrtValueReUse) {
+  // We create a simple model with 2 optional ops - Optional and OptionalGetElement
+  // that don't produce graph outputs and we have logic in the allocation planner to
+  // re-use input OrtValues as output OrtValues for such cases. We test that the model
+  // executes fine with sch logic.
+  // TODO: How to ensure that re-use took place ?
+
```

Comment:
One way is checking the allocation plan (assuming the implementation honours the plan correctly), but to access that you need a Model and to create an InferenceSession via InferenceSessionWrapper with that model. That lets you get SessionState, from which you can call GetExecutionPlan() and look at the allocation_plan member.

Right now OpTester doesn't provide a way to get the Model it builds, although it could make BuildGraph public so it can be called externally.